src/workerd/io/worker.c++

```diff
@@ -216,69 +218,74 @@ void reportStartupError(
-        // We should never get here in production if we've validated scripts before deployment.
-        KJ_LOG(ERROR, "script startup exceeded resource limits", id, ex);
-        kj::throwFatalException(kj::cp(ex));
-      }
-    } else if (catcher.HasCaught()) {
-      js.withinHandleScope([&] {
-        auto exception = catcher.Exception();
-
-        permanentException = js.exceptionToKj(js.v8Ref(exception));
+    KJ_SWITCH_ONEOF(limitErrorOrTime) {
+      KJ_CASE_ONEOF(limitError, kj::Exception) {
+        auto description = jsg::extractTunneledExceptionDescription(limitError.getDescription());
 
+        auto& ex = permanentException.emplace(kj::mv(limitError));
         KJ_IF_SOME(e, errorReporter) {
-          auto limitScope = limitEnforcer.enterLoggingJs(js, maybeLimitError2);
-
-          kj::Vector<kj::String> lines;
-          lines.add(kj::str("Uncaught ", jsg::extractTunneledExceptionDescription(
-              KJ_ASSERT_NONNULL(permanentException).getDescription())));
-          jsg::JsMessage message(catcher.Message());
-          message.addJsStackTrace(js, lines);
-          e.addError(kj::strArray(lines, "\n"));
-
+          e.addError(kj::heapString(description));
         } else KJ_IF_SOME(i, inspector) {
-          auto limitScope = limitEnforcer.enterLoggingJs(js, maybeLimitError2);
-          sendExceptionToInspector(js, *i.get(),
-                                   UncaughtExceptionSource::INTERNAL,
-                                   jsg::JsValue(exception),
-                                   jsg::JsMessage(catcher.Message()));
+          // We want to extend just enough cpu time as is necessary to report the exception
+          // to the inspector here. 10 milliseconds should be more than enough.
+          auto limitScope = limitEnforcer.enterLoggingJs(js, limitErrorOrTime2);
+          jsg::sendExceptionToInspector(js, *i.get(), description);
           // When the inspector is active, we don't want to throw here because then the inspector
           // won't be able to connect and the developer will never know what happened.
         } else {
           // We should never get here in production if we've validated scripts before deployment.
-          kj::Vector<kj::String> lines;
-          jsg::JsMessage message(catcher.Message());
-          message.addJsStackTrace(js, lines);
-          auto trace = kj::strArray(lines, "; ");
-          auto description = KJ_ASSERT_NONNULL(permanentException).getDescription();
-          KJ_LOG(ERROR, "script startup threw exception", id, description, trace);
-          KJ_FAIL_REQUIRE("script startup threw exception");
+          KJ_LOG(ERROR, "script startup exceeded resource limits", id, ex);
+          kj::throwFatalException(kj::cp(ex));
         }
-      });
-    } else {
-      kj::throwFatalException(kj::cp(permanentException.emplace(
-          KJ_EXCEPTION(FAILED, "returned empty handle but didn't throw exception?", id))));
+      }
+      KJ_CASE_ONEOF_DEFAULT {
```

Comment:
Might be better to be specific here and use `KJ_CASE_ONEOF` for the other possible type (same for other KJ_CASE_ONEOF_DEFAULT uses below). This way if the OneOf type gets another variant we'll get a compile-time error rather than this code path just being used implicitly.

Reply:
I changed some of them now.
Some of them make sense as a ONEOF instead of default but the ones who specifically check for an exception shouldn't care if another value is added.
I imagine this as more of a Result type where we changed from `Result<(), Error>` to `Result<Duration, Error>`, we might in the future change to `Result<OneOf<Duration, ??>, Error>`. 
So in some cases it's just more interesting to check whether or not we failed rather than what's the result's success value.